packages/react/dialog/src/Dialog.tsx

```diff
@@ -247,21 +247,26 @@ const DialogContentModal = React.forwardRef((props, forwardedRef) => {
 
 const DialogContentNonModal = React.forwardRef((props, forwardedRef) => {
   const context = useDialogContext(CONTENT_NAME);
-  const isEscapeKeyDownRef = React.useRef(false);
+  const isInteractOutsideRef = React.useRef(false);
+
   return (
     <Portal>
       <DialogContentImpl
         {...props}
         ref={forwardedRef}
         trapFocus={false}
         disableOutsidePointerEvents={false}
-        onCloseAutoFocus={composeEventHandlers(props.onCloseAutoFocus, (event) => {
-          event.preventDefault();
-          if (isEscapeKeyDownRef.current) context.triggerRef.current?.focus();
-        })}
-        onEscapeKeyDown={composeEventHandlers(props.onEscapeKeyDown, () => {
-          isEscapeKeyDownRef.current = true;
-        })}
+        onCloseAutoFocus={composeEventHandlers(
+          props.onCloseAutoFocus,
+          (event) => {
+            if (!event.defaultPrevented && !isInteractOutsideRef.current) {
+              context.triggerRef.current?.focus();
+            }
+            event.preventDefault();
```

Comment:
Man, this is all getting so confusing because of our `checkForDefaultPrevented` thing haha. Next step will really be to do something about this because it's melting my brain 😀

Here for example, it feels weird to see a check for `!event.defaultPrevented` and also have `checkForDefaultPrevented: false` 😀

Also, the placement of our own `event.preventDefault()` bothers me…
- I understand we need to do it in **ALL** cases because either:
  - we don't want to run any focus at all (ie. you've clicked somewhere outside, so natural focus should occur)
  - we want explicit focus to trigger so we prevent the default close auto focus
- So, instinctively I want to put it as the first line at the beginning of the handler, but we can't because we immediately check for `!event.defaultPrevented` (because we need to know if it's been prevented by the user, not us)

As a test I tried to see how it would look like if we didn't use `composeEventHandlers` and ended with this which definitely makes me think we should move away from it soon because I find it way clearer:

```tsx
onCloseAutoFocus={(event) => {
  props.onCloseAutoFocus?.(event);
  const userPrevented = event.defaultPrevented;

  if (!userPrevented) {
    event.preventDefault();
    if (!isInteractOutsideRef.current) context.triggerRef.current?.focus();
  }

  isInteractOutsideRef.current = false;
}}
```

With the current way we end up with a double negative sort of situation, I find it super hard to reason about, especially because for example our own `event.preventDefault()` is run even if the user has already prevented. I know it doesn't make any difference, but it makes it quite confusing when I read it.

What do we all think about this? Do we wait? or do we bite the bullet and start here with this change before updating everywhere later?

Reply:
I'm glad you bring this up because this whole thing often makes my head spin and I find myself wondering whether I'm just dense or if this approach needs adjustment.

In your example, will the consumer setting

```jsx
onCloseAutoFocus={(event) => event.preventDefault()}
```

Revert the auto focus behaviour to the implicit, last focused element thing?

Reply:
No if the user prevent defaults, no focus behaviour will happen because the check is done in `FocusScope` and we pass it down: `onUnmountAutoFocus={onCloseAutoFocus}`.

Reply:
My 2c is that your suggestion makes A LOT more sense and I would prefer to go with it here. I also support incrementally moving away from the utils if we agree that's the right thing to do.

Reply:
Yeah at this point I think I'd also say we should do it incrementally as it seems pointless to have those things use it even though we all feel it's making things worse and harder to understand.

Reply:
Aw man, I totally agree with all of this and **MUCH** prefer your suggestion Benoit. I don't think we need to make huge sweeping codebase changes if we decide to change direction, I think it's fine to start here and avoid `composeEventHandlers` going forward and have a separate task to update legacy stuff.

Reply:
Yay for all agreeing 🎉 
Let's goooo!

Reply:
I've tweaked this for non-modal for now https://github.com/radix-ui/primitives/pull/799/commits/cdc24481073ed22d40c98bc816316ec58a0f719d will update the others in those files separately if we're good.

Question on moving away from the utils, in the common case does this mean we'll just be adopting an inline condition in place?

e.g.

```jsx
onClick={composeEventHandlers(onClick, () => {
  ...
})}
      
// Will become

onClick={(event) => {
  props.onClick?.(event)
  const userPrevented = event.defaultPrevented

  if (!userPrevented) {
   ...
  }
}}
```

Reply:
Yeah although I'm find with just doing:

```jsx
onClick={(event) => {
  props.onClick?.(event)
  if (!event.defaultPrevented) {
   ...
  }
}}
```

It's not necessarily "userPrevented". We could prevent things internally via a capture event or something for example.